PathStream::currentPoint should have a fast path for getting end point immediately

Constellation opened the pull request:
#### 7fcf207fdda17d52fb187617f33064b721f2d556
<pre>
PathStream::currentPoint should have a fast path for getting end point immediately
<a href="https://bugs.webkit.org/show_bug.cgi?id=260652">https://bugs.webkit.org/show_bug.cgi?id=260652</a>
rdar://114371099

Reviewed by Cameron McCormack.

PathStream::currentPoint is computing the end point by traversing all the past segments.
But if we would like to get the end point, in most of cases, we can immediately get that
from the last segment&apos;s information. We are traversing this every time we call `lineTo`
function, and this becomes costly when the path gets many segments.
This patch adds a fast path covering most of cases. We add tryGetEndPointWithoutContext
function, which may return the end point if it can compute it directly without context.
And lineTo function calls this for the last segment and attempt to skip traversing of
segments.

* Source/WebCore/platform/graphics/PathSegment.cpp:
(WebCore::PathSegment::tryGetEndPointWithoutContext const):
* Source/WebCore/platform/graphics/PathSegment.h:
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::PathMoveTo::tryGetEndPointWithoutContext const):
(WebCore::PathLineTo::tryGetEndPointWithoutContext const):
(WebCore::PathQuadCurveTo::tryGetEndPointWithoutContext const):
(WebCore::PathBezierCurveTo::tryGetEndPointWithoutContext const):
(WebCore::PathArcTo::tryGetEndPointWithoutContext const):
(WebCore::PathArc::tryGetEndPointWithoutContext const):
(WebCore::PathEllipse::tryGetEndPointWithoutContext const):
(WebCore::PathEllipseInRect::tryGetEndPointWithoutContext const):
(WebCore::PathRect::tryGetEndPointWithoutContext const):
(WebCore::PathRoundedRect::tryGetEndPointWithoutContext const):
(WebCore::PathDataLine::tryGetEndPointWithoutContext const):
(WebCore::PathDataQuadCurve::tryGetEndPointWithoutContext const):
(WebCore::PathDataBezierCurve::tryGetEndPointWithoutContext const):
(WebCore::PathDataArc::tryGetEndPointWithoutContext const):
(WebCore::PathCloseSubpath::tryGetEndPointWithoutContext const):
* Source/WebCore/platform/graphics/PathSegmentData.h:
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::currentPoint const):

Canonical link: <a href="https://commits.webkit.org/267231@main">https://commits.webkit.org/267231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8e7cb939a72b5d9469af25bb104674e3829e958

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16336 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16728 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17782 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15039 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16205 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19348 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16440 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17475 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16210 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16678 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13665 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18544 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13922 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14481 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21340 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14911 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14654 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17887 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12923 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14480 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18850 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1963 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15065 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam reviewed (approved):
r=me. The way that `applyElements` and `transform` work, which are similar situations where they only work on certain path segment types, is to have a `constexpr` static member on the class indicating whether they support the operation, and then only implement the function on the classes that do support it. I'm not sure in retrospect whether that's a useful pattern versus just returning a `std::optional`, so feel free to leave this patch as is.

Constellation committed:
PathStream::currentPoint should have a fast path for getting end point immediately
https://bugs.webkit.org/show_bug.cgi?id=260652
rdar://114371099

Reviewed by Cameron McCormack.

PathStream::currentPoint is computing the end point by traversing all the past segments.
But if we would like to get the end point, in most of cases, we can immediately get that
from the last segment's information. We are traversing this every time we call `lineTo`
function, and this becomes costly when the path gets many segments.
This patch adds a fast path covering most of cases. We add tryGetEndPointWithoutContext
function, which may return the end point if it can compute it directly without context.
And lineTo function calls this for the last segment and attempt to skip traversing of
segments.

* Source/WebCore/platform/graphics/PathSegment.cpp:
(WebCore::PathSegment::tryGetEndPointWithoutContext const):
* Source/WebCore/platform/graphics/PathSegment.h:
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::PathMoveTo::tryGetEndPointWithoutContext const):
(WebCore::PathLineTo::tryGetEndPointWithoutContext const):
(WebCore::PathQuadCurveTo::tryGetEndPointWithoutContext const):
(WebCore::PathBezierCurveTo::tryGetEndPointWithoutContext const):
(WebCore::PathArcTo::tryGetEndPointWithoutContext const):
(WebCore::PathArc::tryGetEndPointWithoutContext const):
(WebCore::PathEllipse::tryGetEndPointWithoutContext const):
(WebCore::PathEllipseInRect::tryGetEndPointWithoutContext const):
(WebCore::PathRect::tryGetEndPointWithoutContext const):
(WebCore::PathRoundedRect::tryGetEndPointWithoutContext const):
(WebCore::PathDataLine::tryGetEndPointWithoutContext const):
(WebCore::PathDataQuadCurve::tryGetEndPointWithoutContext const):
(WebCore::PathDataBezierCurve::tryGetEndPointWithoutContext const):
(WebCore::PathDataArc::tryGetEndPointWithoutContext const):
(WebCore::PathCloseSubpath::tryGetEndPointWithoutContext const):
* Source/WebCore/platform/graphics/PathSegmentData.h:
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::currentPoint const):

Canonical link: https://commits.webkit.org/267231@main
